[owenyeo] IP

owenyeo opened the pull request:
### ChatBot

> "A wise man once said work smart not hard." - Someone smart, probably

ChatBot frees your mind of having to remember things you need to do. It's,

- Text-based
- Easy to learn and master
- **SUPER** convenient to use!
- Funny and charismatic! 

All you need to do is,

1. download it from [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/522).
2. double-click it.
3. add your tasks.
4. let it **manage your tasks for you** 😄 

And it is FREE!

Features:

- [x]  Managing tasks
- [x]  Keep track of important dates
- [ ]  GUI
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

Look at the charismatic Bot's introduction!
```
/**
     * Prints an introduction.
     */
    public void intro() {
        print(new String[] {"Hello! I am Bobby Wasabi", 
            "What can I do for you today?"});
    }
```

owenyeo committed:
Added Skeleton code that outputs the required output as per source website.
Added functionality to rename chatbot.

owenyeo committed:
Added echoing functionality
Added functionality to read User input
Added ability to exit chat when User Input is "bye"
Provided documentation.
Removed Message subclass due to redundancy.

owenyeo committed:
Added the functionality of listing
Added functionality to add items into list

owenyeo committed:
Added functionality to unmark and mark tasks as done

owenyeo committed:
Added functionality to add ToDo, Event, and Deadline
Added a Command Enum for cleaner code
Added new classes Task, Event, Todo, Deadline

owenyeo committed:
Added ChatBotException, InvalidCommandException, and InvalidDescrptionException.
Added error handling in ChatBot.

owenyeo committed:
Added delete functionality
Added InvalidIndexException for number errors in mark, unmark, and delete.

owenyeo committed:
Added InvalidIndexException
Added saveTasks to save data on a text file.
Added loadTasks for potential future use.

owenyeo committed:
Improved on documentation
Improved formatting of toSaveString in ChatBot.java
Fixed bug where type of task is not correctly shown.
Removed loadFile due to redundancy.

owenyeo committed:
Added functionality to store DateTime objects
Added functionality to reformat DateTime when printing.

owenyeo committed:
Refactored code into different files for more OOP
Added functionality for Storage to load previous lists.
Added SaveFileNotFound exception

owenyeo committed:
Reorganised code into packages.
Added documentation for all new classes.
Fixed bug when marking tasks as done and reloading it in the next session.

owenyeo committed:
Add Gradle support for Chat Bot

Previous version of ChatBot did not have support for Gradle.

Support for gradle is required as we ramp up automated unit testing.
Gradle support is therefore added, with JUnit tests coming in the next commit.
Gradle is used for its ease of use and compatibility with JUnit.

Fixed bugs regarding error messages not printing out.

owenyeo committed:
Add JUnit tests into IP

No automated unit testing was implemented in this project. This led to time inefficiency as I had to
manually test each component one by one.

Therefore, JUnit tests are being implemented to test the various components in this project. Currently,
I am implementing only tests for TaskList and Parser. More will be implemented as required.

Fixed a bug regarding InvalidDescriptionException not being thrown when an empty description is passed through.

owenyeo committed:
Follow coding convention

No coding convention was adhered to. This causes code to be less readable by others.

Thus, from now on the project will adhere to Java Coding Convention. Checkstyle will be implemented and
consistent checks from developer side will continue to ensure this.

owenyeo committed:
Add Find command

There is no find command currently in the chatbot. This makes it hard to look for tasks that
the user may specifically is looking for.

A find command is thus added and is currently working as per developer's testing.

wj331 commented on src/main/java/ChatBot.java:
```diff
@@ -0,0 +1,333 @@
+
+    //Enum Command to make code cleaner and allow for the use of 
+    //switch case statements.
+    private static enum Command {
```
Maybe consider using Commands instead of Command?

wj331 commented on src/main/java/ChatBot.java:
```diff
@@ -0,0 +1,333 @@
+    public void addToList(String taskString, Command command) 
+        throws InvalidDescriptionException {
+        switch(command) {
+            case ADD_TODO:
```
indentation of case should be same as switch i think?

wj331 commented on src/main/java/ChatBot.java:
```diff
@@ -0,0 +1,333 @@
+                    list.add(new Event(eventLabel.trim(), eventParts2[0].trim(), eventParts2[1].trim()));
+                    saveTasks();
+                } catch (IndexOutOfBoundsException e) {
+                    throw new InvalidDescriptionException("Are you stupid? Can you follow instructions?");
```
haha this actually made me laugh. q funny

wj331 commented on src/main/java/ChatBot.java:
```diff
@@ -0,0 +1,333 @@
+     * @param listNum Index of the item of the list to delete.
+     */
+    public void delete(int listNum) {
+        int index = listNum - 1;
```
Perhaps be more descriptive? index of (indexDone)?

wj331 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,51 @@
+ * @author Owen Yeo
+ */
+public class Event extends Task{
+    private LocalDateTime from;
```
Perhaps use a more descriptive name? eg fromTime? from feels abit too general it could be from (location) instead of from (time)?

wj331 reviewed (commented):
LGTM, just some nits on naming etc

owenyeo committed:
Add Checkstyle Support

Project has no checkstyle support, which makes it difficult to check for errors
pertaining to coding convetion.
Thus, checkstyle support has been added to automate checking and ensure better code quality.

zsh-eng commented on src/main/java/chatbot/ChatBot.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public void run() {
```
Perhaps a JavaDoc comment for this method?

zsh-eng commented on src/main/java/chatbot/commands/Command.java:
```diff
@@ -0,0 +1,53 @@
+     * 
+     * @return false
+     */
+    public boolean isExit() {return false;}
```
Should this method follow Egyptian style braces instead?

zsh-eng commented on src/main/java/chatbot/commands/UnmarkItem.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Storage storage, Ui ui) 
+        throws InvalidIndexException {
```
Should this wrapped line be indented by 8 spaces instead?

zsh-eng commented on src/main/java/chatbot/exceptions/SaveFileNotFound.java:
```diff
@@ -0,0 +1,9 @@
+package chatbot.exceptions;
+
```
Perhaps a JavaDoc comment here to describe the class.

zsh-eng commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,77 @@
+            throw new InvalidCommandException("What are you saying? Try again.");
+        }
+
+        switch(command) {
+            case BYE:
+                return new Bye("", CommandType.BYE);
+
+            case DISPLAY_LIST:
+                return new DisplayList("", CommandType.DISPLAY_LIST);
+
+            case MARK:
+                return new MarkItem(inputStrings[1], CommandType.MARK);
+
+            case UNMARK:
+                return new UnmarkItem(inputStrings[1], CommandType.UNMARK);
+
+            case ADD_TODO:
+                return new AddToDo(inputStrings[1], CommandType.ADD_TODO);
+
+            case ADD_DEADLINE:
+                return new AddDeadline(inputStrings[1], CommandType.ADD_DEADLINE);
+            
+            case ADD_EVENT:
+                return new AddEvent(inputStrings[1], CommandType.ADD_EVENT);
+
+            case DELETE:
+                return new DeleteItem(inputStrings[1], CommandType.DELETE);
+
+            case FIND:
+                return new FindTask(inputStrings[1], CommandType.FIND);
+
+            default:
+                throw new InvalidCommandException("Don't be stupid, speak english.");
```
I like how clean this looks. Keep it up!

zsh-eng commented on src/main/java/chatbot/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+        this.originalString = deadline;
+    }
+
+    /**
+     * 
+     */
+    @Override
+    public String toSaveString() {
+        return "D " + super.toSaveString() + " | " + originalString;
+    }
+
+    /**
+     * 
+     */
```
Perhaps a JavaDoc comment here? Since the methods are overriden, I think removing the comment is fine too!

zsh-eng commented on src/main/java/chatbot/ui/Ui.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Prints a goodbye message.
+     */
+    public void bye() {
+        print(new String[] {"Bye. Have a bad day you doofus."});
+    }
```
Perhaps a method name that starts with a verb, for example `printGoodbye` would be more descriptive?

For your reference:
>  Names representing methods must be verbs and written in camelCase.

zsh-eng reviewed (commented):
Overall, some minor changes with naming and comments. LGTM!

owenyeo committed:
Add GUI support for ChatBot

There is no GUI for ChatBot, which is not ideal for User Experience. A UI
was therefore added using JavaFX, and is currently working well as per developer
testing and JUnit tests.

owenyeo committed:
Add FXML Support to GUI

ChatBot has no support for FXML, which makes GUI design clunkier and more tedious.

FXML Support thus has been added to facilitate GUI design. Original functions seem
to be working well as per unit testing and developer tests.

owenyeo committed:
Add Assertion statements

Project does not have any assert statements, meaning that potential issues with logic may go unnoticed.
Adding assertion statements thus will allow for better developer testing and ensure code works as intended.
As a step towards bettering code quality, assertion statements have been added to some classes where errors
tend to happen, e.g. tasklist and tasks with timings.

owenyeo committed:
Improve Code Quality

Code has some minor violations in coding standards throughtout the project.
This leads to lower readability of code and thus makes it hard to debug.

Checkstyle is thus used to detect such errors, and code is screened to correct
any violation detected.

owenyeo committed:
Add functionality to detect duplicates

Chatbot did not recognise and handle duplicate entries of the same task. This
led to clunky UX as users added multiple entries of the same task.

Added functionality to inform users of duplicate entries when adding new tasks, to list
duplicates, and to delete all duplicates if any.

Fixed a bug regarding delete where the wrong number of remaining tasks left is shown.

owenyeo committed:
Add User Guide and Screenshot

owenyeo committed:
Add Screenshot and User Guide

owenyeo committed:
Improve README

README formatting was wrong and looks clunky. Fixed remaining issues.

owenyeo committed:
Fix hard-coding file path

File path is hardcoded into chatBot, causing .jar files to
not save tasks properly. Bug has been fixed.

owenyeo committed:
Fix bug with FXML

App could not run on windows systems due to FXML compatibalitity issues.
Imported packages for Windows and Linux to ensure cross-platform compatibality

Brushed up on some header commands.

owenyeo committed:
Moved README to doc folder

README is outside of docs folder, making it messy.
Moved README to docs folder.
